chalice/package.py

```diff
@@ -594,3 +605,67 @@ def _fixup_deployment_package(self, template, outdir):
                 self._osutils.copy(original_location, new_location)
                 copied = True
             resource['Properties']['CodeUri'] = './deployment.zip'
+
+
+class TemplateMergePostProcessor(TemplatePostProcessor):
+    def process(self, template, config, outdir, chalice_stage_name):
+        # type: (Dict[str, Any], Config, str, str) -> None
+        if not self._should_merge_template(config):
+            return
+        loaded_template = self._load_template_to_merge(config)
+        merger = TemplateDeepMerger()
+        merged = merger.merge(loaded_template, template)
+        template.clear()
+        template.update(merged)
+
+    def _should_merge_template(self, config):
+        # type: (Config) -> bool
+        return config.package_merge_template is not None
+
+    def _load_template_to_merge(self, config):
+        # type: (Config) -> Dict[str, Any]
+        filepath = os.path.abspath(config.package_merge_template)
+        if not self._osutils.file_exists(filepath):
+            raise RuntimeError('Cannot find template file: %s' % filepath)
+        template_data = self._osutils.get_file_contents(filepath, binary=False)
+        try:
+            loaded_template = json.loads(template_data)
+        except ValueError:
+            raise RuntimeError(
+                'Expected %s to be valid JSON template.' % filepath)
+        return loaded_template
+
+
+class TemplateDeepMerger(object):
+    def merge(self, src, dst):
```

Comment:
FWIW, one of the things that made this hard to read was that the `src` and `dst` seemed backwards to me (in terms of their semantics).  Expanding this into the general case of merging a number of templates someone specifies I was expecting that :

* `src` - The base template that chalice generates
* `dst` (or eventually a list of `dst`) - Overrides/overlays that get merged on top of the values that chalice generates.

This seems to make sense to me because you have a base layer, and you overlay additional resources on top of this base layer.  This has the same behavior it would have now in that the users's template will override things that get generated in chalice.

Reply:
I was thinking of it in terms of copying a source on top of a destination. But I agree its confusing. I can switch it to something like chalice_template, extra_template or something much more explicit so you don't need to map src/dst to real concepts when reading it.